### DISCUSSION TITLE
Create windowed app without pythonw

### DISCUSSION THREAD

**Discussion by DelofJ** [pyinstaller/pyinstaller#7787] [Help] [answered]
So here is my issue

I have a small GUI and CLI app in Python
I can run directly, making my UI appearing, or run it in the terminal with some args that do not trigger the UI, making it usable through a terminal

When I run PyInstaller with "--onefile" and "--windowed" on my Mac running Python 3.8.9 (it does not have a pythonw), I get a .app bundle that I can run directly without triggering the terminal while still having my UI, but I can also run the executable inside "Contents/MacOS" within a terminal to use it as a CLI app. I don't know how that happens, but I think it's related to the fact that there is no pythonw on my current installation, but it's really useful

So I also want to bundle my app on my PC running Python 3.11.4 (it has pythonw), I run it with "--onefile" and "--windowed", and it results in an .exe file that I can run directly to have my UI without a terminal, but if I try to run it through an actual terminal, there is no output, it is basically the same behavior as using pythonw to run my script, but I would like to have a behavior similar to the app I have on my Mac

So here is my question, is there a way to bundle a Python app with PyInstaller that can be ran directly without triggering the terminal but also getting an output when it's ran through a terminal ?

Edit: I noticed what I said is a bit confusing, don't hesitate if you would like to have more details or another explanation

**Comment by rokm** [comment, marked as answer]
> So here is my question, is there a way to bundle a Python app with PyInstaller that can be ran directly without triggering the terminal but also getting an output when it's ran through a terminal ?

On Windows, no. 

Console being (automatically) attached and available to the application depends on whether the executable was built [with CONSOLE subsystem or with WINDOWS subsystem](https://learn.microsoft.com/en-us/cpp/build/reference/subsystem-specify-subsystem?view=msvc-170). That's why python on windows has two interpreter variants (`python.exe` and `pythonw.exe`), built with different subsystems. And similarly, PyInstaller on Windows uses two bootloader variants (called `run.exe` and `runw.exe`), depending on whether you use console mode (`--console` or `--nowindowed`) or windowed mode (`--noconsole` or `--windowed`).

If you want a "hybrid" behavior, you will need to fake it - make a console enabled application that detects whether its process owns console or not, and hide it accordingly (for onefile build, it is actually a bit more complicated, because of two processes involved, so you need to detect whether the parent of your application process owns the console). See #7729 for a similar discussion around this, and #7735 for proof-of-concept implementation.

**Reply by DelofJ** [reply]
Well, today I learned, thanks for the help
